### ISSUE TITLE
Hierarchical storage

### ISSUE THREAD

**Issue by marcelgarus** [hivedb/hive#102]
**What's the main motivation?**
In the real world, you often have items that are dependant on each other. A relational database intrinsically models these relations, but even document-based databases like Firestore support hierarchical documents.

**Are there real-world use cases for that?**
Of course, here are just two examples in my current project where this feature would be super useful:
- A file explorer, where the file structure get cached locally. Directory entries depend on other directory entry, so if one directory gets deleted on the server, all the cached files inside it get deleted from the cache too.
- A lecturing course that has some lessons or homework. If the course gets deleted, its lessons should also be deleted from the cache.

**So, what to do?**
I built a simple wrapper package, [hive_cache](https://pub.dev/packages/hive_cache), for this use case, but for now, it's badly documented.
I'd love to see first-party integration of something like hierarchical boxes. You could even emulate hierarchical boxes by saving the relations of items and cleverly deleting them, like proposed in #101.

**Comment by leisim** [comment]
The implementation of the new [`HiveList`](https://github.com/hivedb/hive/blob/dev/hive/lib/src/hive_list.dart) is already finished. I think both of the given examples can be implemented using `HiveList`s.

It also provides a method called `deleteAllFromBox()`. You could just override the `delete()` method of your `HiveObject` and also call `deleteAllFromBox()` on the list. Let's take the file system example:

```dart
@HiveType()
class FSEntity extends HiveObject {
  @HiveField(0)
  String name;

  @HiveField(1)
  HiveList<FSEntity> children;

  @override
  void delete() {
    children.deleteAllFromHive();
    super.delete();
  }
}
```

I always try to keep `Hive` as universal as possible. I think the vast majority of problems can be solved using `HiveList`. What do you think?

**Comment by marcelgarus** [comment]
That's great, I'll be happy to use that. When is this going to be released?

**Comment by marcelgarus** [comment]
Okay great 😄 I'm looking forward to it

**Comment by peekpt** [comment]
@leisim I'm trying to do the "Link" thing but I have no idea what I'm doing 😄

as described on #101 it's a HiveLink thing to hold two key together in order to relationship lists between boxes.

Here's what I got so far:

We have a box of Post, a box of Tag and box of Link

link.dart
```dart
import 'package:hive/hive.dart';

part 'link.g.dart';

@HiveType()
class Link<A extends HiveObject, B extends HiveObject> extends HiveObject {
  @HiveField(0)
  Box<A> boxOne;
  @HiveField(1)
  Box<B> boxTwo;
  @HiveField(3)
  dynamic keyOne;
  @HiveField(4)
  dynamic keyTwo;
 
  Link();
  Link.set(A one, B two) {
    boxOne = one.box;
    boxTwo = two.box;

    keyOne = one.key;
    keyTwo = two.key;
  }

  A get one => boxOne.get(keyOne);
  B get two => boxTwo.get(keyTwo);
}
``` 

post.dart 
```dart
import 'package:hive/hive.dart';

part 'post.g.dart';

@HiveType()
class Post extends HiveObject {
  @HiveField(0)
  String title;
  Post(this.title);
}
```

tag.dart
```dart
import 'package:hive/hive.dart';

part 'tag.g.dart';

@HiveType()
class Tag extends HiveObject {
  @HiveField(0)
  String name;
  Tag(this.name);
}
```

main() 

```dart
var dir = await getApplicationDocumentsDirectory();
  Hive.init(dir.path);

  Hive.registerAdapter(PostAdapter(), 0);
  Hive.registerAdapter(TagAdapter(), 1);
  Hive.registerAdapter(LinkAdapter(), 2);

  var postsBox = await Hive.openBox<Post>('posts');
  var tagsBox = await Hive.openBox<Tag>('tags');
  var links = await Hive.openBox<Link<Post, Tag>>('links');

  await postsBox.clear();
  await tagsBox.clear();
  await links.clear();

  postsBox.add(Post("The great Bananas"));
  postsBox.add(Post("A wonderfull Apple"));
  postsBox.add(Post("Ferrari is good"));

  tagsBox.add(Tag("fruits"));
  tagsBox.add(Tag("cars"));
  tagsBox.add(Tag("animals"));

  var posts = postsBox.values.toList();

  var tags = tagsBox.values.toList();

  var link1 = Link.set(posts[0], tags[0]);
  var link2 = Link.set(posts[1], tags[0]);
  var link3 = Link.set(posts[2], tags[1]);
  links.add(link1);
  links.add(link2);
  links.add(link3);

  // list all posts with tag 1

  links.values.forEach((link) {
    print("link: ${link.one.title} - [${link.two.name}]");
  });
```

console
```bash
I/flutter (21678): link: The great Bananas - [fruits]
I/flutter (21678): link: A wonderfull Apple - [fruits]
I/flutter (21678): link: Ferrari is good - [cars]
```

and I've got this exception, I've no idea:

```
HiveError: Cannot write, unknown type: BoxImpl<Post>. Did you forget to register an adapter?
```

**Comment by leisim** [comment]
The problem is that Hive does not know how to serialize a `Box` object (in your case `boxOne` and `boxTwo`. I would serialize the name instead and create a getter to get the box:

```dart
@HiveType()
class Link<A extends HiveObject, B extends HiveObject> extends HiveObject {
  @HiveField(0)
  String _boxNameOne;
  Box<A> get boxOne => Hive.box<A>(_boxNameOne);

  @HiveField(1)
  String _boxNameTwo;
  Box<B> get boxTwo => Hive.box<B>(_boxNameTwo);

  @HiveField(3)
  dynamic keyOne;

  @HiveField(4)
  dynamic keyTwo;
 
  Link();
  Link.set(A one, B two) {
    _boxNameOne = one.box.name;
    _boxNameTwo = two.box.name;

    keyOne = one.key;
    keyTwo = two.key;
  }

  A get one => boxOne.get(keyOne);
  B get two => boxTwo.get(keyTwo);
}
```

**Comment by peekpt** [comment]
Thank you it worked fine! I also put the parameters private. Also the delete functions. But I'm not sure about the terminology  😄 (one two)
Null test are not implemented yet

not sure how to notify the listed links 😞 If an item get's changed the list will not notify

So far with this method we can List
all Posts with Tags
A Tag with Posts.  (like Hivelist)
A Post with Tags.  (like Hivelist)
All delete functions.
It can be added a bunch filtering helper methods. 

´´´dart
import 'package:hive/hive.dart';

part 'link.g.dart';

@HiveType()
class Link<A extends HiveObject, B extends HiveObject> extends HiveObject {
  @HiveField(0)
  String _boxNameOne;

  @HiveField(1)
  String _boxNameTwo;

  @HiveField(3)
  dynamic _keyOne;

  @HiveField(4)
  dynamic _keyTwo;

  Link();
  Link.set(A one, B two) {
    _boxNameOne = one.box.name;
    _boxNameTwo = two.box.name;

    _keyOne = one.key;
    _keyTwo = two.key;
  }

  A get one => Hive.box<A>(_boxNameOne).get(_keyOne);

  B get two => Hive.box<B>(_boxNameTwo).get(_keyTwo);

  deleteWithOne() {
    one.delete();
    delete();
  }

  deleteWithTwo() {
    two.delete();
    delete();
  }

  deleteAll() {
    one.delete();
    two.delete();
    delete();
  }
}
´´´

**Comment by peekpt** [comment]
@leisim  I'm reading about indexed dbs and relationships and I came across this:

**One-to-many relationship
One-to-many relationship is modeled by keeping ‘one’ reference to objects in ‘many’ side. Reference is simply a primary key, which could be a string or an integer.

This could be a solution for Hivelist

**Comment by leisim** [comment]
> I'm reading about indexed dbs and relationships and I came across this.

Could you maybe share a link where you found this?

**Comment by peekpt** [comment]
http://dev.yathit.com/ydn-db/doc/usage/modeling.html

**Comment by peekpt** [comment]
Regarding one-to-many relationships and the concept of having the reference on the many side:
Instead of having HiveLink with two objects. 
The user should have a way to shortcut objects from other boxes. 

A Basket has many bananas.

If we store the reference (HiveKey) on the many side the user could create a list on the fly by listing all bananas and filter by the reference. That key has a getter HiveKey.value 


_not tested, just for a general idea._
```dart
import 'package:hive/hive.dart';

part 'banana.g.part';

@HiveType()
class Banana extends HiveObject {
  @HiveField(0)
  String name;
  @HiveField(1)
  HiveKey basket;

}

part 'basket.g.part';

@HiveType()
class Basket extends HiveObject {
  @HiveField(0)
  String name;
  
}

@HiveType()
class HiveKey<E extends HiveObject> extends HiveObject {
  @HiveField(0)
  dynamic key;
  @HiveField(1)
  String boxName;

  HiveKey(E obj) : key = obj.key, boxName = obj.box.name;

  E get value => Hive.box<E>(boxName).get(key);

  deleteAll(){
    value.delete();
    delete();
  }
}
```

**Comment by pro100svitlo** [comment]
any update on this?

**Comment by leisim** [comment]
> any update on this?

I made progress but it is still going to take some time...

**Comment by leisim** [comment]
HiveLists are ready and published in `v1.2.0`.
